// react-native-parsed-text/.eslintrc.js
module.exports = {
  rules: {
    indent: [
      2,
      2,
      {
        SwitchCase: 1,
      },
    ],
    quotes: [2, 'single', 'avoid-escape'],
    'comma-dangle': [2, 'always-multiline'],
    'linebreak-style': [2, 'unix'],
    semi: [2, 'always'],
    'space-before-function-paren': [2, 'never'],
    'space-return-throw-case': [2],
    'array-bracket-spacing': [2, 'never'],
    'space-before-blocks': [2],

    'jsx-quotes': [2, 'prefer-single'],
    'key-spacing': [
      2,
      {
        beforeColon: false,
        afterColon: true,
      },
    ],
    'comma-spacing': [
      2,
      {
        before: false,
        after: true,
      },
    ],
    'space-after-keywords': [2, 'always'],
    'block-spacing': [2, 'always'],
    'no-unused-vars': [
      2,
      {
        vars: 'all',
        varsIgnorePattern: 'React',
      },
    ],
    'no-cond-assign': [0],
    eqeqeq: [2, 'allow-null'],
    'react/display-name': 1,
    'react/forbid-prop-types': 1,
    'react/jsx-boolean-value': [2, 'always'],
    'react/jsx-closing-bracket-location': 2,
    'react/jsx-curly-spacing': 1,
    'react/jsx-indent-props': [2, 2],
    'react/jsx-max-props-per-line': 1,
    'react/jsx-no-bind': 1,
    'react/jsx-no-duplicate-props': 1,
    'react/jsx-no-undef': 1,
    'react/jsx-quotes': 1,
    'react/jsx-uses-react': 1,
    'react/jsx-uses-vars': 1,
    'react/no-danger': 1,
    'react/no-did-mount-set-state': 1,
    'react/no-did-update-set-state': 1,
    'react/no-direct-mutation-state': 1,
    'react/no-multi-comp': 1,
    'react/no-set-state': 1,
    'react/no-unknown-property': 1,
    'react/prefer-es6-class': 1,
    'react/prop-types': 1,
    'react/react-in-jsx-scope': 1,
    'react/require-extension': 1,
    'react/self-closing-comp': 1,
    'react/sort-comp': 1,
    'react/wrap-multilines': 1,
  },
  env: {
    es6: true,
    browser: true,
    amd: true,
  },
  globals: {
    module: true,
    describe: true,
    beforeEach: true,
    afterEach: true,
    global: true,
    expect: true,
    it: true,
  },
  extends: 'eslint:recommended',
  ecmaFeatures: {
    jsx: true,
    modules: true,
    experimentalObjectRestSpread: true,
  },
  parser: 'babel-eslint',
  plugins: ['react'],
};
